File: 4. Impossible - Longest Valid Parentheses/LongestValidParentheses.java
import java.util.Stack;

class   {
    public int longestValidParentheses(String s) {
        Stack<Integer> stack = new Stack<>();
        stack.push(-1);
        int longest = 0;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '(') {
                stack.push(i);
            } else if (c == ')') {
                stack.pop();
                if (stack.isEmpty()) {
                    stack.push(i);
                } else {
                    longest = Math.max(longest, i - stack.peek());
                }
            }
        }
        return longest;
    }
}

File: 4. Impossible - Longest Valid Parentheses/LongestValidParentheses.py
'''
My solution to the problem uses a stack to keep track of the indices of the parentheses.
I iterate through the string and when I find an open parenthesis, I add its index to the stack.
When I find a close parenthesis, I pop the last element from the stack and calculate the length of the valid parentheses.
I keep track of the longest valid parentheses and return it at the end.
'''
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        stack = [-1]
        longest =0
        for i in range(len(s)):
            if s[i] == "(":
                stack.append(i)
            elif s[i] == ")":
                stack.pop()
                if len(stack) == 0:
                    stack.append(i)
                elif len(stack) != 0:
                    longest = max(longest, i-stack[-1])
        return longest